Add highlighter task

mcbouslog opened the pull request:
## Package
- lib-classifier

## Linked Issue and/or Talk Post
- towards #4326 
- based on [the PFE Highlighter task](https://github.com/zooniverse/Panoptes-Front-End/tree/master/app/classifier/tasks/highlighter)

## Describe your changes
- adds highlighter task component, model, annotation
- updates text viewer for highlights

## Open Issues
- [x] annotation `end` value incorrect - a51f1413b1ab9949424c8b6f6110818d0b56e9dd
- [x] selection delete button styling - 1a4ea6264b6ca216dc27e43feca083fe36619175

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX?
  - fe-project-branch - https://fe-project-branch.preview.zooniverse.org/projects/markbouslog/digileap-testing-production/classify/workflow/23429
  - local/app-project - https://local.zooniverse.org:3000/projects/markb-panoptes/digileap-testing-staging/classify/workflow/3538
  - local/lib-classifier - https://local.zooniverse.org:8080/?project=1952&workflow=3538
- What user actions should my reviewer step through to review this PR?
  1. highlight text in the text viewer
  2. click one of the highlighter task label buttons in the task area
  3. repeat steps 1-2 as necessary
  4. click Done button in task area 
- Which storybook stories should be reviewed?
  - http://localhost:6006/?path=/story/tasks-highlighter--default
  - http://localhost:6006/?path=/story/subject-viewers-singletextviewer--with-labeled-text
- Are there plans for follow up PR’s to further fix this bug or develop this feature? no

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

## Post-merge
- [ ] This PR adds translations keys to English dictionary(s). See guidance for pulling new keys to Lokalise [here](https://github.com/zooniverse/how-to-zooniverse/blob/master/Translations/lokalise.md#lokalise-and-fem).

eatyourgreens commented on packages/lib-classifier/src/plugins/tasks/experimental/highlighter/models/HighlighterAnnotation.js:
```diff
@@ -0,0 +1,29 @@
+
+const Highlighter = types.model('Highlighter', {
+  taskType: types.literal('highlighter'),
+  value: types.optional(types.array(Highlight), [])
```
```suggestion
  value: types.array(Highlight)
```

Tiny change here: `types.array` implements `types.optional` by default, so you only need to override that if you don't want the default empty array.

eatyourgreens commented on packages/lib-classifier/src/plugins/tasks/experimental/highlighter/models/HighlighterAnnotation.js:
```diff
@@ -0,0 +1,29 @@
+})
+  .views(self => ({
+    get isComplete () {
+      return self.value.length > 0
```
Could a volunteer be shown a page that doesn't have any text labels? If so, this might catch them out by requiring at least one label on each subject.

mcbouslog replied:
The highlighter task isn't required (something to revisit, it prob should have the option to be required), so as is a user can not highlight anything and will be allowed to proceed (Next/Done).

Here's r[elated task model code](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/plugins/tasks/models/Task.js#L36-L38).
Here's a link to the a [highlighter task in the Project Builder](https://master.pfe-preview.zooniverse.org/lab/1952/workflows/3538).

eatyourgreens commented:
I think I'm missing something here. I can't label text selections at all.


https://user-images.githubusercontent.com/59547/231740147-a0707edd-08db-449c-82d1-77119285d1cb.mov

eatyourgreens commented:
It kind of works in Chrome. First thing I've noticed is that the character counts are off by one in the annotations. According to this, 'NEW' has four letters (indexes 0 to 3 in the string; should be 0 to 2.)

<img width="494" alt="Screenshot of a highlighter task annotation in the dev console, with start and end indexes for the text." src="https://user-images.githubusercontent.com/59547/231742222-33afcc7b-64ed-4442-a5ac-56922b69643a.png">

eatyourgreens commented:
I'm not sure if this is a bug as such, more weird UX, but selection fails if I select already selected text as part of a new selection.


https://user-images.githubusercontent.com/59547/231743389-c56fdf71-d305-4488-a7fe-565c42ac01e8.mov

eatyourgreens commented on packages/lib-classifier/src/plugins/tasks/experimental/highlighter/components/HighlighterTask.js:
```diff
@@ -0,0 +1,188 @@
+
+  function handleClick (event, index) {
+    const selection = document.getSelection()
+    selection.rangeCount ? createLabelAnnotation(selection, index) : null
```
I think this check is failing in Firefox. Clicking the buttons does nothing.

mcbouslog replied:
Turns out Firefox wasn't reading `ariaLabel` on a parent node per the `selectableArea` helper as expected, [this commit](https://github.com/zooniverse/front-end-monorepo/pull/4370/commits/da5cb18f467b72fbd8d13a0b93e2f6da0d2ba29d#diff-40a9affb9a1951ca60e6c41e528c5e62aa577536938dc6485973890263ed1294L4) fixes.

mcbouslog replied:
Noting the button to delete a selection needs better styling in Firefox and Safari.

eatyourgreens replied:
Oh yes, if you're working with a HTML element, rather than a React component, you should use `getAttribute` to get the ARIA label.
```js
  const ariaLabel = element.getAttribute('aria-label');
```

eatyourgreens replied:
Also, a GitHub markdown tip: commit hashes are automatically converted to links: eg. da5cb18f467b72fbd8d13a0b93e2f6da0d2ba29d

eatyourgreens commented:
I checked [the original highlighter tool](https://master.pfe-preview.zooniverse.org/projects/eatyourgreens/text-highlighter-demo/classify?env=staging) in Firefox, and that does work. 🤔

mcbouslog commented:
> It kind of works in Chrome. First thing I've noticed is that the character counts are off by one in the annotations. According to this, 'NEW' has four letters (indexes 0 to 3 in the string; should be 0 to 2.)
> 
> <img alt="Screenshot of a highlighter task annotation in the dev console, with start and end indexes for the text." width="494" src="https://user-images.githubusercontent.com/59547/231742222-33afcc7b-64ed-4442-a5ac-56922b69643a.png">

Looking into this now, and the following doesn't change a related fix, but noting I think PFE might also do the same.
![PFE_highlighter_annotation](https://user-images.githubusercontent.com/12118751/231871153-a43c5ef3-2c47-47fd-87dc-8a929e821f1b.png)

mcbouslog commented:
I deleted the [`extraNewLineCharacter` helper](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/classifier/tasks/highlighter/index.jsx#L36-L46) as it's no longer relevant based on the browsers we support. I tested Chrome, Edge, Firefox, and Safari and didn't see the noted extra character when highlighting a word at the end of a line.

I changed the annotation `end` calculation to reflect the last character in a selection as started selecting with an index of 0.

I removed some selection delete button styling. I think Firefox and Safari look a lot better, though Chrome and Edge look a little off, I think the styling looks good enough across all browsers.

eatyourgreens commented:
I don’t know if this is a blocking issue, but the highlighter doesn’t work on iOS any more. 

https://user-images.githubusercontent.com/59547/233260817-1fbce5db-35a1-4aa3-b5a9-0d1d5cf4780f.mp4

eatyourgreens commented:
Looks good in desktop Firefox. Here, I'm tagging text with a combination of trackpad and keys (using the keyboard to modify the initial selection.)

https://user-images.githubusercontent.com/59547/233298713-3231f26f-f71f-4f81-91ec-7680541a88ea.mov

eatyourgreens commented:
Highlighting and tagging text works in Android 12, but not on iOS 16 (tested with a Samsung Galaxy and a iPad on Browserstack. I also tested mobile Safari on my phone.)

Hopefully, the mobile Safari bug is something minor.

mcbouslog commented:
Noting Sean has shown me a bug with a user clicking within the text, but not selecting any text, can click label and create empty selection (and keep clicking to create multiple empty selections) I'll be following up on.

eatyourgreens commented:
>Noting Sean has shown me a bug with a user clicking within the text, but not selecting any text, can click label and create empty selection (and keep clicking to create multiple empty selections) I'll be following up on.

Should the task be disabled, except when there's a selected range in the subject area?

mcbouslog commented:
Noting a few style edits:
- change dark mode "0 Marked" font color to Dark-5, 5c5c5c
- change dark mode highlight button background to Dark-4, 404040
- change dark mode highlight button text to white ffffff

mcbouslog commented:
I created #4498 to follow-up on the the highlighter task not working on iOS.
Thank you Jim for noting and the very helpful screen capture!

mcbouslog commented on packages/lib-classifier/src/plugins/tasks/experimental/highlighter/models/HighlighterAnnotation.js:
```diff
@@ -0,0 +1,29 @@
+  start: types.optional(types.number, 0),
+  end: types.optional(types.number, 0),
+  text: types.optional(types.string, ''),
+  labelInformation: types.optional(HighlighterLabel, { color: '', label: '' })
```
Great questions about aggregation and translations for this task. I'll test out translations and request the research team test it as well. I'm not sure about the research team's exact plans with aggregation, but will note these potential issues.

mcbouslog commented:
> 

Great suggestion! I added `tabIndex` to Selection and `onKeyDown` handler so on backspace or delete it'll delete the selection.

mcbouslog commented:
> I'm thinking of word processor toolbars, where bold and italic buttons aren't clickable unless you've actually selected some text.

Good idea, I'll bring up with Sam and Sean and see what they think! Thanks.

mcbouslog commented:
I'm going to merge this for now, I'll follow-up on the iOS issue and any other issues that come up in further review with subsequent PRs. Thank you @eatyourgreens !
